### PR TITLE
Permutation selection

### PR DESCRIPTION
**Pull request by CongMa13**
Added the permutation instance selection.

- Removed duplicated code in device folder. 
- Removed Query which is not used in solution_registry
- Added selection of instances based on lengths of input tensor
    - selection function is hard code now. 

The work flow 

image

### CODE REVIEW

File: library/src/permutation/device/hiptensor_permutation_scale_instances.hpp [hunk 1]
```diff
@@ -41,106 +49,76 @@ namespace ck
...
-                            else if constexpr(is_same_v<InDataTypeTuple, ck::Tuple<F16>>
-                                              && is_same_v<OutDataTypeTuple, ck::Tuple<F16>>)
-                            {
-                                add_device_permute_scale_6d_f16_instances(op_ptrs);
-                            }
-                        }
-                        return op_ptrs;
+                        std::unordered_map<hiptensor::Uid, std::unique_ptr<DeviceOp>> opPtrs;
+                        // clang-format off
+                        addInstance<256, 64, 64, 4, 4, ck::Sequence<0, 1>, ck::Sequence<4>, ck::Sequence<4>>(opPtrs);
```
Comment (CongMa13) [inline library/src/permutation/device/hiptensor_permutation_scale_instances.hpp:115]: Will add the instances with best performance in another PR when the performance test is done.

File: library/src/permutation/device/CMakeLists.txt [hunk 2]
```diff
@@ -26,11 +26,7 @@
 
  set(CK_PERMUTATION_INSTANCE_SOURCES
      ${CMAKE_CURRENT_SOURCE_DIR}/hiptensor_permutation_scale_instances.hpp
```
Comment (cgmillette) [inline library/src/permutation/device/CMakeLists.txt:28]: Not sure why there are hpp files in source list?

File: library/src/permutation/device/instance_params.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                                                  scale,
+                                                  numDim,
+                                                  blockSize,
+                                                  m0PerBlock,
+                                                  m1PerBlock,
+                                                  m0PerThread,
+                                                  m1PerThread,
+                                                  threadClusterArrangeOrder.first,
+                                                  threadClusterArrangeOrder.second,
+                                                  scalarPerVectorSeq,
```
Comment (cgmillette) [inline library/src/permutation/device/instance_params.cpp]: Do both in and out have to be the same? Or is it interesting to have a combinatorial set for both in and out?

File: library/src/permutation/device/CMakeLists.txt (hunk 2 above)
Comment (CongMa13) [reply library/src/permutation/device/CMakeLists.txt:28]: It is not necessary to include header files. But it seems to be helpful to IDE. I'm not quite sure about the IDE. Do you prefer to keep it or not?

---------------------

In CMake, it's not strictly necessary to add header files to the source file list of a target. CMake primarily needs the source files to compile the project. However, including header files can be beneficial for IDE integration and for ensuring that they are visible in the project structure.

Here's a quick example of how you might include header files:

```cmake
add_executable(MyApp
    src/main.cpp
    src/other.cpp
    include/myheader.h
)
```

This way, your IDE can recognize and display the header files properly.

File: library/src/permutation/device/instance_params.cpp (hunk 3 above)
Comment (CongMa13) [reply library/src/permutation/device/instance_params.cpp]: Yeah, hiptensor supports combinations. 

image

cuTensor supports even more. 

|typeA | typeB | typeScalar
-- | -- | --
CUDA_R_16F | CUDA_R_16F | CUDA_R_16F
CUDA_R_16F | CUDA_R_16F | CUDA_R_32F
CUDA_R_16F | CUDA_R_32F | CUDA_R_32F
CUDA_R_32F | CUDA_R_16F | CUDA_R_32F
CUDA_R_16BF | CUDA_R_16BF | CUDA_R_16BF
CUDA_R_16BF | CUDA_R_16BF | CUDA_R_32F
CUDA_R_32F | CUDA_R_32F | CUDA_R_32F
CUDA_R_64F | CUDA_R_64F | CUDA_R_64F
CUDA_R_32F | CUDA_R_64F | CUDA_R_64F
CUDA_R_64F | CUDA_R_32F | CUDA_R_64F
CUDA_C_32F | CUDA_C_32F | CUDA_C_32F
CUDA_C_64F | CUDA_C_64F | CUDA_C_64F
CUDA_C_32F | CUDA_C_64F | CUDA_C_64F
CUDA_C_64F | CUDA_C_32F | CUDA_C_64F

File: library/src/permutation/device/CMakeLists.txt (hunk 2 above)
Comment (cgmillette) [reply library/src/permutation/device/CMakeLists.txt:28]: Can keep them if helpful

**Review by cgmillette — DISMISSED** [review]

File: test/utils.hpp [hunk 4]
```diff
@@ -308,7 +311,7 @@ std::pair<bool, double> compareEqualLaunchKernel(DDataType*             deviceD,
                        0,
                        deviceD,
                        hostD,
-                       d_relativeError,
+                       static_cast<double*>(relativeErrorPtr.get()),
                        elementsD);
     CHECK_HIP_ERROR(hipEventRecord(syncEvent));
     CHECK_HIP_ERROR(hipEventSynchronize(syncEvent));
```
Comment (cgmillette) [inline test/utils.hpp:317]: missing destroy for this sync event

Comment (CongMa13) [reply test/utils.hpp:317]: added hipEventDestroy

**Review by cgmillette — APPROVED** [review]
LGTM

**Pull request merged**